python/lib/externalrequests/dmod/externalrequests/duplex/handler.py

```diff
@@ -0,0 +1,1104 @@
+        """
+        recognized_types: typing.Set[typing.Type[FieldedMessage]] = set()
+
+        for message_type in self.__source_handler_routing:
+            recognized_types.add(message_type)
+
+        for message_type in self.__target_handler_routing:
+            recognized_types.add(message_type)
+
+        return recognized_types
+
+    def parse_socket_input(self, socket_input: typing.Union[str, bytes]) -> typing.Union[str, FieldedMessage, bytes]:
+        """
+        Convert the passed socket data into either a concrete message or just a string or byte
+
+        Args:
+            socket_input: The data that arrived through a socket
+
+        Returns:
+            A deserialized version of the socket data
+        """
+        try:
+            parsed_input = None
+
+            if self._recognized_message_types:
+                # Try and actually deserialize the input into a dictionary for further identification
+                parsed_input = json.loads(socket_input)
+
+            if parsed_input:
+                # Try to convert the JSON into a valid message type
+                for recognized_type in self._recognized_message_types:
+                    message = recognized_type.factory_init_from_deserialized_json(parsed_input)
+                    if message:
+                        return message
+        except:
+            # The input could not be parsed into JSON to check for a concrete message
+            pass
+
+        if isinstance(socket_input, bytes):
+            try:
+                socket_input = socket_input.decode()
+            except:
+                # The data could not be decoded, which is fine - this may happen if pure binary data is passed
+                pass
+
+        return socket_input
+
+    async def listen_to_messages(
+        self,
+        source: WebSocketCommonProtocol,
+        target: WebSocketCommonProtocol,
+        path: str,
+        typed_handlers: typing.Mapping[typing.Type[FieldedMessage], typing.Iterable[TYPED_MESSAGE_HANDLER]],
+        untyped_handlers: typing.Iterable[MESSAGE_HANDLER],
+        *args,
+        **kwargs
+    ) -> Response:
+        """
+        Listen for messages from the server
+
+        Args:
+            source: The socket connection that provides messages
+            target: The socket connection that does not provide messages
+            path: The path to this handler on the server
+            typed_handlers: Message handlers for messages that can be deserialized into a FieldedMessage
+            untyped_handlers: Message handlers for messages that can't be deserialized into Fielded messages
+            *args:
+            **kwargs:
+
+        Returns:
+            A response reporting on the success of this operation
+        """
+        # Make sure that the source is a websocket protocol and open so that messages may be correctly listened to
+        if not isinstance(source, WebSocketCommonProtocol):
+            raise ValueError(
```

Comment:
Thoughts on changing this to a `TypeError`?

Reply:
There's a bunch of places where TypeError makes more sense. Changing it now.

Reply:
Tanks!